File: sysrap/tests/S4MTRandGaussQTest.py
#!/usr/bin/env python

import numpy as np
SIZE = np.array([1280, 720])
import matplotlib.pyplot as mp


def test_transformQuick():
    a = np.load(os.path.expandvars("$FOLD/test_transformQuick.npy"))
    fig, ax = mp.subplots(figsize=SIZE/100.)
    ax.scatter( a[:,0], a[:,1], s=0.1 )
    fig.show() 
    return a 


def test_transformQuick_vs_njuffa_erfcinvf():
    a = np.load(os.path.expandvars("$FOLD/test_transformQuick_vs_njuffa_erfcinvf.npy"))
    fig, ax = mp.subplots(figsize=SIZE/100.)
    ax.scatter( a[:,0], a[:,1], s=0.1, label="transformQuick" )
    ax.scatter( a[:,0], a[:,2], s=0.1, label="njuffa_erfcinvf" )
    ax.legend()
    fig.show() 
    return a


def test_shoot():
    a = np.load(os.path.expandvars("$FOLD/test_shoot.npy"))
    fig, ax = mp.subplots(figsize=SIZE/100.)
    ax.hist(a, bins=100)
    fig.show() 
    return a



if __name__ == '__main__':
    #a = test_shoot()
    a = test_transformQuick_vs_njuffa_erfcinvf()

pass

File: sysrap/tests/S4MTRandGaussQTest.cc
#include "S4MTRandGaussQ.h"
#include "NP.hh"
#include "njuffa_erfcinvf.h"

void test_transformQuick()
{
    int N = 1000 ; 
    NP* a = NP::Make<double>( N, 2 ); 
    double* aa = a->values<double>() ;

    for(int i=0 ; i < N ; i++)
    {
        double r = double(i)/double(N-1)  ;   
        double v = S4MTRandGaussQ::transformQuick(r) ; 
        aa[i*2+0] = r ; 
        aa[i*2+1] = v ; 
    }
    a->save("$FOLD/test_transformQuick.npy"); 
}

void test_transformQuick_vs_njuffa_erfcinvf()
{
    int N = 1000 ;

    int ni = N ; 
    int nj = 3 ; 
 
    NP* a = NP::Make<double>( ni, nj ); 
    double* aa = a->values<double>() ;

    for(int i=0 ; i < ni ; i++)
    {
        double r = double(i)/double(N-1)  ;   
        aa[i*nj+0] = r ; 
        aa[i*nj+1] = S4MTRandGaussQ::transformQuick(r)  ; 
        aa[i*nj+2] = -sqrtf(2.f)*njuffa_erfcinvf(r*2.f) ; 
        // argument to erfcinvf needs to be in range 0->2 
        // then have to scale to match S4MTRandGaussQ::transformQuick
    }
    a->save("$FOLD/test_transformQuick_vs_njuffa_erfcinvf.npy"); 
}


void test_shoot()
{
    int N = 1000000 ; 
    NP* a = NP::Make<double>( N ); 
    double* aa = a->values<double>() ;

    double mean = 0.0 ; 
    double stdDev = 0.1 ; 

    for(int i=0 ; i < N ; i++) aa[i] = S4MTRandGaussQ::shoot(mean, stdDev) ; 
    a->save("$FOLD/test_shoot.npy"); 
}


int main()
{
    //test_transformQuick(); 
    //test_shoot(); 

    test_transformQuick_vs_njuffa_erfcinvf(); 

    return 0 ; 
}
